### ISSUE TITLE
Proposal: Allow DevNull to be used in redirect_stdout

### ISSUE THREAD

**Issue by sbromberger** [JuliaLang/julia#22879]
Following up on https://github.com/JuliaLang/julia/issues/12050#issuecomment-316493469,

I would like to propose that we change `redirect_stdout` to allow this:
```
julia> redirect_stdout(DevNull)
ERROR: MethodError: no method matching redirect_stdout(::Base.DevNullStream)
Closest candidates are:
  redirect_stdout() at stream.jl:1034
  redirect_stdout(::Union{Base.LibuvStream, IOStream}) at stream.jl:1028
  redirect_stdout(::Function, ::Any) at stream.jl:1082
```

Right now,

`DevNull <: DevNullStream <: IO <: Any`,
`open("/tmp/foo", "w") <: IOStream <: IO <: Any`,
`STDOUT <: TTY <: LibuvStream <: IO <: Any`

The latter two are allowed with `redirect_stdout` per the second method listed, but `DevNull` isn't. 

(Is there a reason that `redirect_stdout` can't accept `IO`? That would resolve this issue and make things more consistent.)

**Comment by yuyichao** [comment]
> Is there a reason that redirect_stdout can't accept IO

Yes, it doesn't work in general.

**Comment by sbromberger** [comment]
> Yes, it doesn't work in general.

Under what circumstances would an `IO` object not be a suitable candidate for `redirect_stdout`?

**Comment by yuyichao** [comment]
> Under what circumstances would an IO object not be a suitable candidate for redirect_stdout?

Basically anything that's not a real file.

**Comment by sbromberger** [comment]
> Basically anything that's not a real file.

But that's not accurate: `TCPStream`, `STDOUT`, and `STDERR` aren't "real files" and they currently work with `redirect_stdout`. I'm afraid I'm not following you here.

**Comment by yuyichao** [comment]
Real file refer to a real file for the OS. All of the above are real files.

**Comment by sbromberger** [comment]
```
base64.jl:mutable struct Base64EncodePipe <: IO
base64.jl:mutable struct Base64DecodePipe <: IO
boot.jl:mutable struct CoreSTDOUT <: IO end
boot.jl:mutable struct CoreSTDERR <: IO end
coreio.jl:struct DevNullStream <: IO end
distributed/clusterserialize.jl:mutable struct ClusterSerializer{I<:IO} <: AbstractSerializer
iobuffer.jl:mutable struct AbstractIOBuffer{T<:AbstractVector{UInt8}} <: IO
iostream.jl:mutable struct IOStream <: IO
mmap.jl:mutable struct Anonymous <: IO
serialize.jl:mutable struct SerializationState{I<:IO} <: AbstractSerializer
show.jl:struct IOContext{IO_t <: IO} <: AbstractPipe
```

Which of these represent "non-real" files that would not be suitable candidates for `redirect_stdout`?

Honestly, it feels like I'm pulling teeth here.

**Comment by vtjnash** [comment]
All of them would not be?

**Comment by sbromberger** [comment]
@vtjnash yet at least some of them are accepted, no? That is,

`open("/tmp/foo", "w") <: IOStream <: IO <: Any`

is a counterexample.

**Comment by JeffBezanson** [comment]
What Yichao means is that the object needs to have a kernel-level file descriptor. The reason is that this redirects all output from the current process, even data written by C code. So we could potentially support `DevNull` by opening `/dev/null`, and passing that file handle.

**Comment by sbromberger** [comment]
@JeffBezanson thanks - that's what I was looking for. In fact, `redirect_stdout(open(/dev/null, "w"))` works, but I can't imagine that's platform-independent. How does `DevNull` do its magic?

**Comment by yuyichao** [comment]
Also

```
iostream.jl:mutable struct IOStream <: IO
```

is not a "non-real" file.

**Comment by JeffBezanson** [comment]
`DevNull` just implements the julia I/O interface with functions that do (almost) nothing. IIUC, it is then supported as a special case by the process code, which passes on values to libuv that cause it to do a portable equivalent of opening /dev/null.

**Comment by sbromberger** [comment]
@JeffBezanson thanks. Since this issue has gone a bit off the rails, I'll close it out and reopen another one with the specific enhancement request.

**Comment by yuyichao** [comment]
I don't think you should open a new one.

**Comment by yuyichao** [comment]
The thread here is useful to prevent anyone else from wanting to redirect to `::IO`.

**Comment by sbromberger** [comment]
ref: #22881

**Comment by JeffBezanson** [comment]
Dup of #18268, but this issue now has more info it seems.

**Comment by StefanKarpinski** [comment]
Is there anything that implementing this would break or is this a pure feature? Asking for a friend named "1.0 milestone".

**Comment by jpfairbanks** [comment]
Ah so the thing about `redirect_stdout` is that it is capturing the stdout of `ccall`s so it needs to have a filehandle. 

I guess from an interface perspective we want everyone who is writing code that does IO to have versions that take an IO object. Otherwise a bad pattern could develop where people capture the output of libraries by redirecting stdout. The caller should pass a buffer for the output if they want to capture it. and libraries need to accept a buffer (defaulting to stdout) to allow that.

**Comment by sbromberger** [comment]
> Is there anything that implementing this would break or is this a pure feature? 

(Not the question you asked, but...)

From my perspective, there's nothing particularly urgent about this, and its implementation probably merits some further discussion. For example, perhaps it makes sense for `DevNull` to be backed by a file handle (if we can do it portably), given that it shares its name with a UNIX system file. This way it can be used for anything that requires a "real file".

If not, then perhaps it should be renamed so as not to cause this sort of confusion.

**Comment by StefanKarpinski** [comment]
No reason non-breaking things can't happen before 1.0 if someone implements them, I'm just asking because we're trying to be very on top of the administrativa of milestone issues and such.

**Comment by Keno** [comment]
Can't think of a reason this would be breaking.

**Comment by Seelengrab** [comment]
The specific issue of redirecting to DevNull has been fixed in #36146, so can be closed?
